rclcpp/test/rclcpp/test_generic_client.cpp

```diff
@@ -0,0 +1,499 @@
+// Copyright 2023 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <string>
+#include <memory>
+#include <utility>
+
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/serialization.hpp"
+
+#include "rcl_interfaces/srv/list_parameters.hpp"
+
+#include "../mocking_utils/patch.hpp"
+#include "../utils/rclcpp_gtest_macros.hpp"
+
+#include "test_msgs/srv/empty.hpp"
+
+using namespace std::chrono_literals;
+
+// All tests are from test_client
+
+class TestGenericClient : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+
+  void SetUp()
+  {
+    node = std::make_shared<rclcpp::Node>("test_node", "/ns");
+  }
+
+  void TearDown()
+  {
+    node.reset();
+  }
+
+  rclcpp::Node::SharedPtr node;
+};
+
+class TestGenericClientSub : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+
+  void SetUp()
+  {
+    node = std::make_shared<rclcpp::Node>("test_node", "/ns");
+    subnode = node->create_sub_node("sub_ns");
+  }
+  void TearDown()
+  {
+    node.reset();
+  }
+
+  rclcpp::Node::SharedPtr node;
+  rclcpp::Node::SharedPtr subnode;
+};
+
+/*
+   Testing client construction and destruction.
+ */
+TEST_F(TestGenericClient, construction_and_destruction) {
+  using rcl_interfaces::srv::ListParameters;
+  {
+    auto client = node->create_generic_client("test_service", "test_msgs/srv/Empty");
+  }
+
+  {
+    ASSERT_THROW(
+    {
+      auto client = node->create_generic_client("invalid_test_service?", "test_msgs/srv/Empty");
+    }, rclcpp::exceptions::InvalidServiceNameError);
+  }
+
+  {
+    ASSERT_THROW(
+    {
+      auto client = node->create_generic_client("test_service", "test_msgs/srv/InvalidType");
+    }, std::runtime_error);
+  }
+}
+
+TEST_F(TestGenericClient, construction_with_free_function) {
+  {
+    auto client = rclcpp::create_generic_client(
+      node->get_node_base_interface(),
+      node->get_node_graph_interface(),
+      node->get_node_services_interface(),
+      "test_service",
+      "test_msgs/srv/Empty",
+      rmw_qos_profile_services_default,
```

Comment:
Maybe we would want to use `rmw_qos_profile_services_default` in `rclcpp`?

Reply:
Should use rclcpp::ServicesQoS().